### DISCUSSION TITLE
Evaluating suitability for fish larvae simulation

### DISCUSSION THREAD

**Discussion by maeckha** [CliMA/Oceananigans.jl#1440] [General]
For a thesis project about the migration and dispersal of fish larvae in the sea I am looking for CFD tools that allow a newcomer in this field (like myself) the simulation of agent based models (each agent represents a fish) in laminar flows and later in more realistic turbulences. I chose the following steps as milestones:

- First, as a means for validation, I will show a lid driven cavity flow (Oceananigans already does this in an example).
- After this I will combine this with an agent based model where an agent will represent a larvae. At that state the agent should follow simple rules, like drift in the direction of a laminar flow.
- If that works, I will implement a continuity equation to simulate the odor which comes from the reef and which is supposed to help the larvae to find back to the home reef.
- Next, putting n agents into the water and see, if the individuals are behaving differently when put into a swarm. 

I am planning to keep this in 2D since doing it in 3D would probably be to difficult given the limited time (3 months). I will not try to simulate the exact ocean currents or the bottom of the ocean.

Are you capable of assessing the suitability of Oceananigans for those steps? Is it possible to deliver those steps? 

For the Agent based modelling I will be working with the [Agents.jl](https://juliadynamics.github.io/Agents.jl/stable/) package probably, since it seems well documented.
Since I have found Oceananigans, which is also a CFD tool, I was wondering what you think the differences are between that and [WaterLily.jl](https://github.com/weymouth/WaterLily.jl)? Given my description of the project above, would you tend to use either of them?
Thank you!

**Comment by ali-ramadhan** [comment]
Hi @maeckha! Glad you're considering Oceananigans.jl.

I think you have a pretty good plan in place and it seems quite doable within 3 months I believe. I guess you're trying to couple an agent-based larvae model to an ocean model?

Oceananigans.jl should have the features you need I think: mainly Lagrangian particle tracking to advect the larvae around and arbitrary forced passive tracers for the odor from the reef?

I've actually been meaning to set up an example of an agent-based model coupled to Oceananigans.jl (simulating interacting "microbes" in the ocean). It's not fully complete yet but might give you some ideas of how to use Lagrangian particles as agents: https://github.com/CliMA/Oceananigans.jl/pull/1244

> If that works, I will implement a continuity equation to simulate the odor which comes from the reef and which is supposed to help the larvae to find back to the home reef.

What is the continuity equation for the odor? It should be easy to advect the odor around as a passive tracer but just curious what the equation looks like.

> Since I have found Oceananigans, which is also a CFD tool, I was wondering what you think the differences are between that and WaterLily.jl? Given my description of the project above, would you tend to use either of them?

I'm not super familiar with WaterLily.jl but it's definitely a very nice package (and a great README)!

I think WaterLily.jl has a more mature immersed boundary implementation if you need to have complex boundaries and is auto-diff friendly if you need to do some optimization/inference. @weymouth might be able to elaborate!

Oceananigans.jl has native support for Lagrangian particle tracking and might have more mature diagnostics (e.g. if you need to compute vorticity or other statistics) and more mature output writers (among other quality of life features). The examples in the docs should give a pretty good of the current set of mature features. E.g. this example shows how to add in a passive tracer for plankton and evolve it according to some (continuity?) equation: https://clima.github.io/OceananigansDocumentation/stable/generated/convecting_plankton/

EDIT: Just converted the issue to a GitHub discussion as it seemed more appropriate.

**Reply by maeckha** [reply]
Thank you for the evaluation. For the last week I was making myself familiar with the tool and its packages. For example I have been running some examples and the validation tests to understand how the component work together.
As soon as I will have worked out the continuity equation I will share it with you.

**Reply by ali-ramadhan** [reply]
Awesome! I think some of the validation experiments might be out of date (we need to update them and add them to CI) but the examples should be up to date.

Feel free to post back with any questions/issues!

**Reply by maeckha** [reply]
Do you mean out of date as in not working anymore? Because for the lid_driven_cavity I am getting this Error message:

```
`ERROR: LoadError: ArgumentError: length(size) must be 2.

Stacktrace:
 [1] validate_tupled_argument(::Tuple{Int64,Int64,Int64}, ::Type{T} where T, ::String, ::Int64; greater_than::Int64) at /home/meck/Documents/Bachelor Arbeit/Repos/Oceananigans.jl/src/Grids/input_validation.jl:24
 [2] validate_tupled_argument(::Tuple{Int64,Int64,Int64}, ::Type{T} where T, ::String, ::Int64) at /home/meck/Documents/Bachelor Arbeit/Repos/Oceananigans.jl/src/Grids/input_validation.jl:24
 [3] validate_size(::Type{T} where T, ::Type{T} where T, ::Type{T} where T, ::Tuple{Int64,Int64,Int64}) at /home/meck/Documents/Bachelor Arbeit/Repos/Oceananigans.jl/src/Grids/input_validation.jl:48
 [4] RegularRectilinearGrid(::Type{T} where T; size::Tuple{Int64,Int64,Int64}, x::Tuple{Int64,Int64}, y::Tuple{Int64,Int64}, z::Tuple{Int64,Int64}, extent::Nothing, topology::Tuple{DataType,DataType,DataType}, halo::Nothing) at /home/meck/Documents/Bachelor Arbeit/Repos/Oceananigans.jl/src/Grids/regular_rectilinear_grid.jl:161
 [5] simulate_lid_driven_cavity(; Re::Int64, N::Int64, end_time::Int64) at /home/meck/Documents/Bachelor Arbeit/Repos/Oceananigans.jl/validation/lid_driven_cavity/lid_driven_cavity.jl:15
 [6] top-level scope at /home/meck/Documents/Bachelor Arbeit/Repos/Oceananigans.jl/validation/lid_driven_cavity/lid_driven_cavity.jl:90
in expression starting at /home/meck/Documents/Bachelor Arbeit/Repos/Oceananigans.jl/validation/lid_driven_cavity/lid_driven_cavity.jl:90`
```
I already tried changing `size` at the mentioned spots.
Is this error due to something on my side?

**Reply by ali-ramadhan** [reply]
Yes we're constantly improving the user interface so some functions change name or the arguments change causing scripts to break. Unfortunately we haven't added the lid-driven cavity to our tests so the script has gotten a bit old and rusty (could definitely too).

But you shouldn't have to modify any internal code, just the `lid_driven_cavity.jl` script.

If you're interested in updating and running it, it would be very much appreciated if you could open a pull request!

We're happy to help with updating the script too.
